ingot-types/src/ip.rs

```diff
@@ -156,54 +184,102 @@ impl Ipv6Addr {
         }
     }
 
+    /// Returns an eight element 16-bit array representation of the address.
+    ///
+    /// This is taken from the core `Ipv6Addr` implementation.
+    #[inline]
+    pub const fn segments(&self) -> [u16; 8] {
+        // All elements in `self.octets` must be big endian.
+        // SAFETY: `[u8; 16]` is always safe to transmute to `[u16; 8]`.
+        let [a, b, c, d, e, f, g, h] = unsafe {
+            core::mem::transmute::<[u8; 16], [u16; 8]>(self.octets())
+        };
+        // We want native endian u16
+        [
+            u16::from_be(a),
+            u16::from_be(b),
+            u16::from_be(c),
+            u16::from_be(d),
+            u16::from_be(e),
+            u16::from_be(f),
+            u16::from_be(g),
+            u16::from_be(h),
+        ]
+    }
+
+    /// Private function to convert to a `core::net::Ipv6Addr`
+    /// in a const context as `From` implementations are not
+    /// yet allowed in const contexts.
+    #[inline]
+    const fn into_core(self) -> core::net::Ipv6Addr {
```

Comment:
I think I'd leave a note here that this can be simplified once [rust#131360](https://github.com/rust-lang/rust/issues/131360) stabilises.